// react-native-notifier/lib/module/Notifier.js
import React from 'react';
import { Animated, View, Platform } from 'react-native';
import { PanGestureHandler, State, TouchableWithoutFeedback } from 'react-native-gesture-handler';
import styles from "./Notifier.styles.js";
import { Notification as NotificationComponent } from "./components/index.js";
import { DEFAULT_ANIMATION_DURATION, DEFAULT_DURATION, MAX_TRANSLATE_Y, MIN_TRANSLATE_Y, SWIPE_ANIMATION_DURATION, SWIPE_PIXELS_TO_CLOSE, DEFAULT_SWIPE_ENABLED, DEFAULT_COMPONENT_HEIGHT } from "./constants.js";
import { FullWindowOverlay } from "./components/FullWindowOverlay.js";
import { jsx as _jsx } from "react/jsx-runtime";
export const Notifier = {
  showNotification: () => {},
  hideNotification: () => {},
  clearQueue: () => {}
};
export class NotifierRoot extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      Component: NotificationComponent,
      swipeEnabled: DEFAULT_SWIPE_ENABLED,
      componentProps: {}
    };
    this.isShown = false;
    this.isHiding = false;
    this.hideTimer = null;
    this.showParams = null;
    this.callStack = [];
    this.hiddenComponentValue = -DEFAULT_COMPONENT_HEIGHT;
    this.translateY = new Animated.Value(MIN_TRANSLATE_Y);
    this.translateYInterpolated = this.translateY.interpolate({
      inputRange: [MIN_TRANSLATE_Y, MAX_TRANSLATE_Y],
      outputRange: [MIN_TRANSLATE_Y, MAX_TRANSLATE_Y],
      extrapolate: 'clamp'
    });
    this.onGestureEvent = Animated.event([{
      nativeEvent: {
        translationY: this.translateY
      }
    }], {
      useNativeDriver: true
    });
    this.onPress = this.onPress.bind(this);
    this.onHandlerStateChange = this.onHandlerStateChange.bind(this);
    this.onLayout = this.onLayout.bind(this);
    this.showNotification = this.showNotification.bind(this);
    this.hideNotification = this.hideNotification.bind(this);
    this.clearQueue = this.clearQueue.bind(this);
    if (!props.omitGlobalMethodsHookup) {
      Notifier.showNotification = this.showNotification;
      Notifier.hideNotification = this.hideNotification;
      Notifier.clearQueue = this.clearQueue;
    }
  }
  componentWillUnmount() {
    clearTimeout(this.hideTimer);
  }
  hideNotification(callback) {
    if (!this.isShown || this.isHiding) {
      return;
    }
    Animated.timing(this.translateY, {
      toValue: this.hiddenComponentValue,
      easing: this.showParams?.hideEasing ?? this.showParams?.easing,
      duration: this.showParams?.hideAnimationDuration ?? this.showParams?.animationDuration ?? DEFAULT_ANIMATION_DURATION,
      useNativeDriver: true
    }).start(result => {
      this.onHidden();
      callback?.(result);
    });
    this.onStartHiding();
  }
  showNotification(functionParams) {
    const {
      // Remove "omitGlobalMethodsHookup" prop as it is only utilized within the constructor and is redundant elsewhere.
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      omitGlobalMethodsHookup,
      // Remove "useRNScreensOverlay" and "rnScreensOverlayViewStyle" as it is only used in the render
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      useRNScreensOverlay,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      rnScreensOverlayViewStyle,
      ...props
    } = this.props;
    const params = {
      ...props,
      ...functionParams,
      componentProps: {
        ...props?.componentProps,
        ...functionParams?.componentProps
      }
    };
    if (this.isShown) {
      switch (params.queueMode) {
        case 'standby':
          {
            this.callStack.push(params);
            break;
          }
        case 'next':
          {
            this.callStack.unshift(params);
            break;
          }
        case 'immediate':
          {
            this.callStack.unshift(params);
            this.hideNotification();
            break;
          }
        default:
          {
            this.callStack = [params];
            this.hideNotification();
            break;
          }
      }
      return;
    }
    const {
      title,
      description,
      swipeEnabled,
      Component,
      componentProps,
      translucentStatusBar,
      containerStyle,
      containerProps,
      onShown,
      ...restParams
    } = params;
    this.setState({
      title,
      description,
      Component: Component ?? NotificationComponent,
      swipeEnabled: swipeEnabled ?? DEFAULT_SWIPE_ENABLED,
      componentProps: componentProps,
      translucentStatusBar,
      containerStyle,
      containerProps
    });
    this.showParams = restParams;
    this.isShown = true;
    this.setHideTimer();
    this.translateY.setValue(-DEFAULT_COMPONENT_HEIGHT);
    Animated.timing(this.translateY, {
      toValue: MAX_TRANSLATE_Y,
      easing: this.showParams?.showEasing ?? this.showParams?.easing,
      duration: this.showParams?.showAnimationDuration ?? this.showParams?.animationDuration ?? DEFAULT_ANIMATION_DURATION,
      useNativeDriver: true
    }).start(onShown);
  }
  clearQueue(hideDisplayedNotification) {
    this.callStack = [];
    if (hideDisplayedNotification) {
      this.hideNotification();
    }
  }
  setHideTimer() {
    const {
      duration = DEFAULT_DURATION
    } = this.showParams ?? {};
    clearTimeout(this.hideTimer);
    if (duration && !isNaN(duration)) {
      this.hideTimer = setTimeout(this.hideNotification, duration);
    }
  }
  onStartHiding() {
    this.showParams?.onStartHiding?.();
    this.isHiding = true;
    clearTimeout(this.hideTimer);
  }
  onHidden() {
    this.showParams?.onHidden?.();
    this.isShown = false;
    this.isHiding = false;
    this.showParams = null;
    this.translateY.setValue(MIN_TRANSLATE_Y);
    const nextNotification = this.callStack.shift();
    if (nextNotification) {
      this.showNotification(nextNotification);
    }
  }
  onHandlerStateChange({
    nativeEvent
  }) {
    if (nativeEvent.state === State.ACTIVE) {
      clearTimeout(this.hideTimer);
    }
    if (nativeEvent.oldState !== State.ACTIVE) {
      return;
    }
    this.setHideTimer();
    const swipePixelsToClose = -(this.showParams?.swipePixelsToClose ?? SWIPE_PIXELS_TO_CLOSE);
    const isSwipedOut = nativeEvent.translationY < swipePixelsToClose;
    Animated.timing(this.translateY, {
      toValue: isSwipedOut ? this.hiddenComponentValue : MAX_TRANSLATE_Y,
      easing: this.showParams?.swipeEasing,
      duration: this.showParams?.swipeAnimationDuration ?? SWIPE_ANIMATION_DURATION,
      useNativeDriver: true
    }).start(() => {
      if (isSwipedOut) {
        this.onHidden();
      }
    });
    if (isSwipedOut) {
      this.onStartHiding();
    }
  }
  onPress() {
    this.showParams?.onPress?.();
    if (this.showParams?.hideOnPress !== false) {
      this.hideNotification();
    }
  }
  onLayout({
    nativeEvent
  }) {
    const heightWithMargin = nativeEvent.layout.height + 50;
    this.hiddenComponentValue = -Math.max(heightWithMargin, DEFAULT_COMPONENT_HEIGHT);
  }
  render() {
    const {
      useRNScreensOverlay,
      rnScreensOverlayViewStyle
    } = this.props;
    const {
      title,
      description,
      swipeEnabled,
      Component,
      componentProps,
      translucentStatusBar,
      containerStyle,
      containerProps
    } = this.state;
    const additionalContainerStyle = typeof containerStyle === 'function' ? containerStyle(this.translateY) : containerStyle;
    return /*#__PURE__*/_jsx(FullWindowOverlay, {
      useOverlay: useRNScreensOverlay,
      viewStyle: rnScreensOverlayViewStyle,
      children: /*#__PURE__*/_jsx(PanGestureHandler, {
        enabled: swipeEnabled,
        onGestureEvent: this.onGestureEvent,
        onHandlerStateChange: this.onHandlerStateChange,
        children: /*#__PURE__*/_jsx(Animated.View, {
          ...containerProps,
          style: [styles.container, {
            transform: [{
              translateY: this.translateYInterpolated
            }]
          }, additionalContainerStyle],
          children: /*#__PURE__*/_jsx(TouchableWithoutFeedback, {
            onPress: this.onPress,
            children: /*#__PURE__*/_jsx(View, {
              onLayout: this.onLayout,
              style: Platform.OS === 'android' && translucentStatusBar ? styles.translucentStatusBarPadding : undefined,
              children: /*#__PURE__*/_jsx(Component, {
                title: title,
                description: description,
                ...componentProps
              })
            })
          })
        })
      })
    });
  }
}
