// react-native-notifier/lib/module/components/Alert.js
import React from 'react';
import { StyleSheet, View, Text, SafeAreaView } from 'react-native';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const bgColors = {
  error: '#E03737',
  warn: '#FFAC00',
  info: '#007BFF',
  success: '#00B104'
};
const s = StyleSheet.create({
  container: {
    margin: 10
  },
  title: {
    fontWeight: 'bold',
    textAlign: 'center',
    fontSize: 15,
    lineHeight: 22,
    color: 'white'
  },
  description: {
    textAlign: 'center',
    fontSize: 14,
    lineHeight: 22,
    color: 'white'
  }
});
const AlertComponent = ({
  title,
  titleStyle,
  description,
  descriptionStyle,
  alertType = 'success',
  backgroundColor,
  textColor,
  ContainerComponent,
  maxTitleLines,
  maxDescriptionLines
}) => {
  const Container = ContainerComponent ?? SafeAreaView;
  const textStyle = textColor ? {
    color: textColor
  } : null;
  return /*#__PURE__*/_jsx(Container, {
    style: {
      backgroundColor: backgroundColor || bgColors[alertType]
    },
    children: /*#__PURE__*/_jsxs(View, {
      style: s.container,
      children: [!!title && /*#__PURE__*/_jsx(Text, {
        style: [s.title, textStyle, titleStyle],
        numberOfLines: maxTitleLines,
        children: title
      }), !!description && /*#__PURE__*/_jsx(Text, {
        style: [s.description, textStyle, descriptionStyle],
        numberOfLines: maxDescriptionLines,
        children: description
      })]
    })
  });
};
export default AlertComponent;
