// react-native-notifier/lib/commonjs/components/Notification.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const s = _reactNative.StyleSheet.create({
  container: {
    shadowColor: '#000000',
    shadowOffset: {
      width: 0,
      height: 2
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 10,
    backgroundColor: '#ffffff',
    borderRadius: 5,
    marginTop: 10,
    marginHorizontal: 10,
    paddingVertical: 10,
    paddingHorizontal: 10,
    flexDirection: 'row',
    alignItems: 'center'
  },
  image: {
    alignItems: 'center',
    marginRight: 5,
    borderRadius: 5,
    height: 45,
    width: 45
  },
  content: {
    flex: 1,
    marginHorizontal: 5
  },
  title: {
    fontWeight: 'bold',
    fontSize: 15,
    lineHeight: 25,
    color: '#007BFF'
  },
  description: {
    fontSize: 14,
    lineHeight: 20,
    color: '#444444'
  }
});
const NotificationComponent = ({
  title,
  titleStyle,
  description,
  descriptionStyle,
  imageSource,
  imageStyle,
  ContainerComponent,
  maxTitleLines,
  maxDescriptionLines,
  containerStyle
}) => {
  const Container = ContainerComponent ?? _reactNative.SafeAreaView;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Container, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
      style: [s.container, containerStyle],
      children: [!!imageSource && /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Image, {
        style: [s.image, imageStyle],
        source: imageSource
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
        style: s.content,
        children: [!!title && /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
          style: [s.title, titleStyle],
          numberOfLines: maxTitleLines,
          children: title
        }), !!description && /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Text, {
          style: [s.description, descriptionStyle],
          numberOfLines: maxDescriptionLines,
          children: description
        })]
      })]
    })
  });
};
var _default = exports.default = NotificationComponent;
